### PR TITLE
Bug fixes and decimal number support addition

### PR DESCRIPTION
**Pull request by vigan-abd**

### CODE REVIEW

File: test/bn.test.js [hunk 1]
```diff
@@ -166,5 +166,114 @@ module.exports = () => {
         validateBN.bind(null, '-0.33')
       ).not.to.throw()
     })
+
+    it("safeSumBN - it should fail on non number values", () => {
```
Comment (robertkowalski) [inline test/bn.test.js]: style: we use single quotes

File: package.json [hunk 2]
```diff
@@ -28,7 +27,8 @@
     "lint": "./node_modules/.bin/standard"
   },
   "dependencies": {
-    "bignumber.js": "^9.0.0"
+    "bignumber.js": "^9.0.0",
+    "decimal.js": "^10.2.0"
```
Comment (robertkowalski) [inline package.json]: no need for decimal js, you can use bignumber.js

File: src/bn.js [hunk 3]
```diff
@@ -10,27 +10,92 @@ const BN = () => BigNumber
 /**
  * @param {string|number|BigNumber} value
  */
-const nBN = (value) => new BigNumber(value)
+const nBN = (value) => {
+  if (Number.isNaN(value) || value === 'NaN') throw new Error('ERR_NUM_NAN')
+  if (typeof value !== 'string' && typeof value !== 'number' &&
+    !(value instanceof BigNumber)) throw new Error('ERR_NUM_NAN')
+  if (value === 'Infinity' || value === '-Infinity' ||
```
Comment (robertkowalski) [inline src/bn.js:17]: very nice!

File: src/bn.js [hunk 4]
```diff
@@ -10,27 +10,92 @@ const BN = () => BigNumber
...
   if (opts.allowZero === false && n.isZero()) throw new Error('ERR_NUM_ZERO')
   if (opts.allowNegative === false && n.isNegative()) throw new Error('ERR_NUM_NOT_POS')
   if (opts.allowDecimals === false && /e|\./.test(n.toString())) throw new Error('ERR_NUM_WITH_PRECISION')
 }
 
+/**
+ * @param {...(string|number|BigNumber)} values
+ * @returns {BigNumber}
+ */
+const safeSumBN = (...values) => {
```
Comment (robertkowalski) [inline src/bn.js]: not sure if we need the additional wrappers for additons, substractions and multiplications, they add possible error sources

**Review by robertkowalski — CHANGES_REQUESTED** [review]
very nice code, enjoyed reading it

probably we can slim it down a bit

File: src/number.js [hunk 5]
```diff
@@ -0,0 +1,15 @@
+'use strict'
+
+const { nBN } = require('./bn')
+
+const safeParseInt = (value) => {
+  const num = parseInt(
+    /^0x/.test(value.toString()) ? value : nBN(value).toNumber(),
```
Comment (robertkowalski) [inline src/number.js]: i don't think we need support for hex or octal

Comment (vigan-abd) [reply src/number.js]: removed the function

File: src/bn.js (hunk 4 above)
Comment (vigan-abd) [reply src/bn.js]: removed the function

File: src/vwap.js [hunk 6]
```diff
@@ -18,26 +23,50 @@ const VWAP = (values) => {
...
-  let x = nBN(0)
-  let y = nBN(0)
+  let sum = nBN(0)
+  weightKeys.forEach(key => {
+    const value = weights[key]
+    validateBN(weights[key], { allowNegative: false })
+    sum = sum.plus(value)
+  })
+
+  if (sum.toFixed() !== '1') throw new Error('ERR_INVALID_WEIGHT_CONF')
```
Comment (robertkowalski) [inline src/vwap.js:47]: very nice! thx

**Review by robertkowalski — APPROVED** [review]

**Pull request merged**